### ISSUE TITLE
Update Smarty template engine to latest working version

### ISSUE THREAD

**Issue by jonyo** [geodesicsolutions-community/geocore-community#22]
As reported in the forums, starting with `3.1.28` the internals may be changing and breaking the GeoCore code.

* Look into what it will take to upgrade to latest version.  Maybe all it needs are minor changes, if so do that.
* If it requires major rewrite of how the geocore templates load, at the very least get it updated to `3.1.27`.
* Whichever version used, if possible, use `composer` to manage the library and remove it from being hard-coded in the source code.

**Comment by jonyo** [comment]
I'm going with "full upgrade" for PHP8 compatibility.  Note to self:  ongoing list of things to test:

- [ ] Listing Checkout (uses `{include file="tset:find:{$template['g_type']}:{$template['g_resource']}:{$template['file']}"}` and not sure if that would actually work...
- [ ] add more as I come to them

**Comment by vicos59** [comment]
Will the existing template files still function, or will they too need updates?

**Comment by jonyo** [comment]
"So far" I have not had to touch any of the main templates, just the system and module templates.  That said, I am not done yet so it may turn out to need changes somewhere.  Any changes that are needed we'll have full instructions on how to do it.

**Comment by jonyo** [comment]
@vicos59 - So far I have not had to touch any of the main templates, just the system and module templates.  That said, I am not done yet so it may turn out to need changes somewhere.

**Comment by vicos59** [comment]
@jonyo Can I just download the latest src/ and copy it over top of my Beta-3? Do I need to run upgrade script?

Anyway, if I can update my Beta-3, I'll start testing this in depth for you.

**Comment by jonyo** [comment]
@vicos59  Yeah, you should be able to copy on top, first I need to finish the changes though.  Oh but you will need to delete all the files in templates_c/ folder.  I turned on all addons and am trying to test, I already found one PHP8 bug I made a separate issue for to fix that isn't template related.

**Comment by vicos59** [comment]
@jonyo OK, I'll standby, just let me know when it is good to grab a copy.

**Comment by jonyo** [comment]
@vicos59 Actually I take it back, you can't just dump `src/` - you have to run `composer install` in the base folder too or you won't have any of the files in `src/vendor/`.

By the way, the changes will include a way to test locally using docker and PHP 8.1 😄

**Comment by vicos59** [comment]
@jonyo Well, my plan was to install this on my test domain on my production server. So.... I need to test it there so I can be sure the real upgrade from 7.5.1 is going to work for my production classifieds. I have zero experience with composer and do not even know if it is available on my server.

**Comment by vicos59** [comment]
Any chance you could just build me a Beta-4?

**Comment by jonyo** [comment]
Sure, I'm going to test a few more places to make sure everything is working at least in PHP 7 then I'll build beta 4.

It may be more buggy than past betas due to the smarty upgrade.

I don't know when I'll have a chance but with any luck, in the next day or so since I have tomorrow off.  (If I'm not too busy spending family time 😄 )

**Comment by vicos59** [comment]
Awesome! Thank you! I finished the update/migration Sunday morning of that site I was working on. On Monday I can see about getting Docker working so testing future versions past Beta-4 won't be an issue.

**Comment by jonyo** [comment]
@vicos59 Ok new release is up for beta 4.

> On Monday I can see about getting Docker working so testing future versions past Beta-4 won't be an issue.

I was just about to ask about docker progress, good to hear!  I hope we are able to pave the way to get some instructions that make "onboarding" go much more smoothly for anyone wanting to try it.

**Comment by vicos59** [comment]
@jonyo So, as far as testing... Is it OK to upgrade and test using PHP 7.4 first, then switch to PHP 8.0. I assume we are still trying to make this at least 7.4 friendly?

**Comment by jonyo** [comment]
@vicos59 
> Is it OK to upgrade and test using PHP 7.4 first, then switch to PHP 8.0. I assume we are still trying to make this at least 7.4 friendly?

Yes test in 7.4 and report any issues.  Priority should be to make sure everything works in PHP 7.4 first, once that is covered, start hunting and fixing PHP 8 issues.

If you do find problems I hope I can see "why I missed it" and find a way to fix the same type of problem across the board, that way we can fix even the fringe areas that may not get tested directly.  This code base has a ton of features, it is not easy to test every single one but hopefully we can get the majority covered.

**Comment by vicos59** [comment]
I'll start a new discussion thread "Beta-3-->Beta-4 upgrade test; Php 7.4 & 8.0" where I will post my notes as I go. When finished we can collate everything and create new issues or whatever is best.
